### DISCUSSION TITLE
What is the best way to identify throttling

### DISCUSSION THREAD

**Discussion by tomkerkhove** [Azure/azure-iot-sdk-csharp#2051] [Q&A]
What is the best way to identify throttling other than checking for the exceptions? Are there response headers, or other ways to keep track of that?

For example, we have an assumption that we are throttled on the IP level and want to see how we can identify this.

**Comment by drwill-ms** [comment]
Hi Tom -

For context, are we talking about device or service SDK? Hub? Which transport protocol are you using?

Thanks,
David

**Reply by tomkerkhove** [reply]
Actually both for device and service in .NET.

**Comment by drwill-ms** [comment]
There are many ways to look at throttling. The most straight-forward one is maximum number of operations per unit of time. Hub defines many additional throttling conditions and limits, so I'll try to include as many of them as possible. You can read more about all [hub quotas and throttling](https://aka.ms/iothubthrottling).

In all cases, the [IsTransient](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.client.exceptions.iothubexception.istransient?view=azure-dotnet#Microsoft_Azure_Devices_Client_Exceptions_IotHubException_IsTransient) property can be checked if the client should attempt retry, or not.

## Service client

For the service SDK, there are several clients: `RegistryManager`, `ServiceClient`, `JobsClient`, and `DigitalTwinsClient`. Transport protocol can influence the kind of errors returned, so be aware of the [table](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/readme.md#iot-hub-service-sdk) that shows which clients and operations happen over HTTP and AMQP.

When you are being throttled in a data-plane SDK, it comes back as an error on the response and is surfaced to the SDK user as an exception. Some of these errors are documented by the service and [here](https://docs.microsoft.com/azure/iot-hub/iot-hub-troubleshoot-error-429001-throttlingexception) is one example for ThrottlingException with an HTTP response code of 429 with a sub code of 001.

- For any operation, if it exceeds the number of operations per second allowed, you could see a [ThrottlingException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.throttlingexception?view=azure-dotnet).
- For any operation throttled over AMQP, you could see a [IotHubThrottledException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.iothubthrottledexception?view=azure-dotnet).
- When quota has been exceeded, you could see a [QuotaExceededException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.quotaexceededexception?view=azure-dotnet).
- Although unlikely, it is possible the service is having trouble managing load. The reported error we'd see in that case is [ServerBusyException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.serverbusyexception?view=azure-dotnet).
- For import/export jobs, you could get a unique throttling exception of [JobQuotaExceededException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.jobquotaexceededexception?view=azure-dotnet) for # of active jobs allowed, given the size of the hub.
- When sending cloud-to-device messages, if the device is not processing them fast enough, you could see [DeviceMaximumQueueDepthExceededException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.devicemaximumqueuedepthexceededexception?view=azure-dotnet).
- If the cloud-to-device message exceeds a maximum threshold, you could see [MessageTooLargeException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.messagetoolargeexception?view=azure-dotnet).
- When doing bulk device registry operations that exceed the published batch limit, one can get [TooManyDevicesException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.toomanydevicesexception?view=azure-dotnet).
- When deploying more than 50 modules to an Edge device, one can get [TooManyModulesOnDeviceException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.toomanymodulesondeviceexception?view=azure-dotnet).
- If the Azure subscription is disabled due to spending limits, you could see [IotHubSuspendedException](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.common.exceptions.iothubsuspendedexception?view=azure-dotnet). This is one that would not be retried automatically, but you'd be unlikely to see it if your subscription is a normal type.

## Device client

For a device SDK, depending on which protocol is used, it may simply be disconnected, or may get an exception. The device client has built-in retry policy, and if left to the default, it will retry indefinitely. However, if the retry policy is overridden, it may be necessary to handle exceptions from throttling. Many of the same exceptions apply on the device. You can see the list [here](https://docs.microsoft.com/dotnet/api/microsoft.azure.devices.client.exceptions?view=azure-dotnet).

We have a good [code sample](https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/master/iot-hub/Samples/device/DeviceReconnectionSample) that describes how to author a device/module app. Please take a look at that to understand how to handle connection events.

Does this help?
